File: agent/internal/metrics/metrics.go
package metrics

import (
	"time"

	"github.com/shirou/gopsutil/v4/cpu"
	"github.com/shirou/gopsutil/v4/disk"
	"github.com/shirou/gopsutil/v4/host"
	"github.com/shirou/gopsutil/v4/load"
	"github.com/shirou/gopsutil/v4/mem"
	"github.com/shirou/gopsutil/v4/net"
)

// SystemInfo holds basic system information
type SystemInfo struct {
	Hostname        string `json:"hostname"`
	Platform        string `json:"platform"`
	PlatformVersion string `json:"platform_version"`
	Architecture    string `json:"architecture"`
	Uptime          uint64 `json:"uptime"` // seconds
	BootTime        uint64 `json:"boot_time"`
	KernelVersion   string `json:"kernel_version"`
}

// CPUMetrics holds CPU usage data
type CPUMetrics struct {
	UsagePercent float64 `json:"usage_percent"` // overall CPU usage
	CoreCount    int     `json:"core_count"`
	Load1        float64 `json:"load_1"`  // 1-minute load average
	Load5        float64 `json:"load_5"`  // 5-minute load average
	Load15       float64 `json:"load_15"` // 15-minute load average
}

// MemoryMetrics holds memory usage data
type MemoryMetrics struct {
	Total       uint64  `json:"total"`
	Used        uint64  `json:"used"`
	Available   uint64  `json:"available"`
	UsedPercent float64 `json:"used_percent"`
	SwapTotal   uint64  `json:"swap_total"`
	SwapUsed    uint64  `json:"swap_used"`
}

// DiskPartition holds info about a single disk partition
type DiskPartition struct {
	Device      string  `json:"device"`
	Mountpoint  string  `json:"mountpoint"`
	Fstype      string  `json:"fstype"`
	Total       uint64  `json:"total"`
	Used        uint64  `json:"used"`
	Free        uint64  `json:"free"`
	UsedPercent float64 `json:"used_percent"`
}

// DiskMetrics holds all disk partition data
type DiskMetrics struct {
	Partitions []DiskPartition `json:"partitions"`
}

// NetworkInterface holds per-interface network stats
type NetworkInterface struct {
	Name        string `json:"name"`
	BytesSent   uint64 `json:"bytes_sent"`
	BytesRecv   uint64 `json:"bytes_recv"`
	PacketsSent uint64 `json:"packets_sent"`
	PacketsRecv uint64 `json:"packets_recv"`
}

// NetworkMetrics holds network I/O data
type NetworkMetrics struct {
	Interfaces []NetworkInterface `json:"interfaces"`
}

// AllMetrics is the top-level response structure
type AllMetrics struct {
	Timestamp int64          `json:"timestamp"`
	System    SystemInfo     `json:"system"`
	CPU       CPUMetrics     `json:"cpu"`
	Memory    MemoryMetrics  `json:"memory"`
	Disk      DiskMetrics    `json:"disk"`
	Network   NetworkMetrics `json:"network"`
}

// Collect gathers all system metrics and returns them
func Collect() (*AllMetrics, error) {
	m := &AllMetrics{
		Timestamp: time.Now().Unix(),
	}

	// System info
	if hi, err := host.Info(); err == nil {
		m.System = SystemInfo{
			Hostname:        hi.Hostname,
			Platform:        hi.Platform,
			PlatformVersion: hi.PlatformVersion,
			Architecture:    hi.KernelArch,
			Uptime:          hi.Uptime,
			BootTime:        hi.BootTime,
			KernelVersion:   hi.KernelVersion,
		}
	}

	// CPU
	if percents, err := cpu.Percent(time.Second, false); err == nil && len(percents) > 0 {
		m.CPU.UsagePercent = percents[0]
	}
	if counts, err := cpu.Counts(true); err == nil {
		m.CPU.CoreCount = counts
	}
	if avg, err := load.Avg(); err == nil {
		m.CPU.Load1 = avg.Load1
		m.CPU.Load5 = avg.Load5
		m.CPU.Load15 = avg.Load15
	}

	// Memory
	if vm, err := mem.VirtualMemory(); err == nil {
		m.Memory.Total = vm.Total
		m.Memory.Used = vm.Used
		m.Memory.Available = vm.Available
		m.Memory.UsedPercent = vm.UsedPercent
	}
	if sm, err := mem.SwapMemory(); err == nil {
		m.Memory.SwapTotal = sm.Total
		m.Memory.SwapUsed = sm.Used
	}

	// Disk
	if parts, err := disk.Partitions(false); err == nil {
		for _, p := range parts {
			if usage, err := disk.Usage(p.Mountpoint); err == nil {
				m.Disk.Partitions = append(m.Disk.Partitions, DiskPartition{
					Device:      p.Device,
					Mountpoint:  p.Mountpoint,
					Fstype:      p.Fstype,
					Total:       usage.Total,
					Used:        usage.Used,
					Free:        usage.Free,
					UsedPercent: usage.UsedPercent,
				})
			}
		}
	}

	// Network
	if counters, err := net.IOCounters(true); err == nil {
		for _, c := range counters {
			m.Network.Interfaces = append(m.Network.Interfaces, NetworkInterface{
				Name:        c.Name,
				BytesSent:   c.BytesSent,
				BytesRecv:   c.BytesRecv,
				PacketsSent: c.PacketsSent,
				PacketsRecv: c.PacketsRecv,
			})
		}
	}

	return m, nil
}

File: app/lib/models/metrics.dart
/// System information from the agent
class SystemInfo {
  final String hostname;
  final String platform;
  final String platformVersion;
  final String architecture;
  final int uptime; // seconds
  final int bootTime;

  SystemInfo({
    required this.hostname,
    required this.platform,
    required this.platformVersion,
    required this.architecture,
    required this.uptime,
    required this.bootTime,
  });

  factory SystemInfo.fromJson(Map<String, dynamic> json) => SystemInfo(
    hostname: json['hostname'] as String? ?? '',
    platform: json['platform'] as String? ?? '',
    platformVersion: json['platform_version'] as String? ?? '',
    architecture: json['architecture'] as String? ?? '',
    uptime: json['uptime'] as int? ?? 0,
    bootTime: json['boot_time'] as int? ?? 0,
  );

  /// Uptime formatted as "Xd Yh Zm"
  String get uptimeFormatted {
    final d = uptime ~/ 86400;
    final h = (uptime % 86400) ~/ 3600;
    final m = (uptime % 3600) ~/ 60;
    if (d > 0) return '${d}d ${h}h ${m}m';
    if (h > 0) return '${h}h ${m}m';
    return '${m}m';
  }
}

/// CPU metrics
class CpuMetrics {
  final double usagePercent;
  final int coreCount;
  final double load1;
  final double load5;
  final double load15;

  CpuMetrics({
    required this.usagePercent,
    required this.coreCount,
    required this.load1,
    required this.load5,
    required this.load15,
  });

  factory CpuMetrics.fromJson(Map<String, dynamic> json) => CpuMetrics(
    usagePercent: (json['usage_percent'] as num?)?.toDouble() ?? 0,
    coreCount: json['core_count'] as int? ?? 1,
    load1: (json['load_1'] as num?)?.toDouble() ?? 0,
    load5: (json['load_5'] as num?)?.toDouble() ?? 0,
    load15: (json['load_15'] as num?)?.toDouble() ?? 0,
  );
}

/// Memory metrics
class MemoryMetrics {
  final int total;
  final int used;
  final int available;
  final double usedPercent;
  final int swapTotal;
  final int swapUsed;

  MemoryMetrics({
    required this.total,
    required this.used,
    required this.available,
    required this.usedPercent,
    required this.swapTotal,
    required this.swapUsed,
  });

  factory MemoryMetrics.fromJson(Map<String, dynamic> json) => MemoryMetrics(
    total: json['total'] as int? ?? 0,
    used: json['used'] as int? ?? 0,
    available: json['available'] as int? ?? 0,
    usedPercent: (json['used_percent'] as num?)?.toDouble() ?? 0,
    swapTotal: json['swap_total'] as int? ?? 0,
    swapUsed: json['swap_used'] as int? ?? 0,
  );

  /// Format bytes to human-readable string
  static String formatBytes(int bytes) {
    if (bytes < 1024) return '$bytes B';
    if (bytes < 1048576) return '${(bytes / 1024).toStringAsFixed(1)} KB';
    if (bytes < 1073741824) return '${(bytes / 1048576).toStringAsFixed(1)} MB';
    return '${(bytes / 1073741824).toStringAsFixed(1)} GB';
  }

  String get totalFormatted => formatBytes(total);
  String get usedFormatted => formatBytes(used);
  String get availableFormatted => formatBytes(available);
}

/// Disk partition info
class DiskPartition {
  final String device;
  final String mountpoint;
  final String fstype;
  final int total;
  final int used;
  final int free;
  final double usedPercent;

  DiskPartition({
    required this.device,
    required this.mountpoint,
    required this.fstype,
    required this.total,
    required this.used,
    required this.free,
    required this.usedPercent,
  });

  factory DiskPartition.fromJson(Map<String, dynamic> json) => DiskPartition(
    device: json['device'] as String? ?? '',
    mountpoint: json['mountpoint'] as String? ?? '',
    fstype: json['fstype'] as String? ?? '',
    total: json['total'] as int? ?? 0,
    used: json['used'] as int? ?? 0,
    free: json['free'] as int? ?? 0,
    usedPercent: (json['used_percent'] as num?)?.toDouble() ?? 0,
  );

  String get totalFormatted => MemoryMetrics.formatBytes(total);
  String get usedFormatted => MemoryMetrics.formatBytes(used);
  String get freeFormatted => MemoryMetrics.formatBytes(free);
}

/// Disk metrics (all partitions)
class DiskMetrics {
  final List<DiskPartition> partitions;

  DiskMetrics({required this.partitions});

  factory DiskMetrics.fromJson(Map<String, dynamic> json) {
    final parts = (json['partitions'] as List<dynamic>?)
        ?.map((e) => DiskPartition.fromJson(e as Map<String, dynamic>))
        .where((p) => p.total > 0 && p.mountpoint != '/dev') // Filter out virtual/empty
        .toList() ?? [];
    return DiskMetrics(partitions: parts);
  }
}

/// Network interface stats
class NetworkInterface {
  final String name;
  final int bytesSent;
  final int bytesRecv;

  NetworkInterface({
    required this.name,
    required this.bytesSent,
    required this.bytesRecv,
  });

  factory NetworkInterface.fromJson(Map<String, dynamic> json) => NetworkInterface(
    name: json['name'] as String? ?? '',
    bytesSent: json['bytes_sent'] as int? ?? 0,
    bytesRecv: json['bytes_recv'] as int? ?? 0,
  );

  String get sentFormatted => MemoryMetrics.formatBytes(bytesSent);
  String get recvFormatted => MemoryMetrics.formatBytes(bytesRecv);
}

/// Network metrics (all interfaces)
class NetworkMetrics {
  final List<NetworkInterface> interfaces;

  NetworkMetrics({required this.interfaces});

  factory NetworkMetrics.fromJson(Map<String, dynamic> json) {
    final ifaces = (json['interfaces'] as List<dynamic>?)
        ?.map((e) => NetworkInterface.fromJson(e as Map<String, dynamic>))
        .where((i) => i.name != 'lo0' && (i.bytesSent > 0 || i.bytesRecv > 0))
        .toList() ?? [];
    return NetworkMetrics(interfaces: ifaces);
  }
}

/// Complete metrics snapshot
class MetricsSnapshot {
  final int timestamp;
  final SystemInfo system;
  final CpuMetrics cpu;
  final MemoryMetrics memory;
  final DiskMetrics disk;
  final NetworkMetrics network;

  MetricsSnapshot({
    required this.timestamp,
    required this.system,
    required this.cpu,
    required this.memory,
    required this.disk,
    required this.network,
  });

  factory MetricsSnapshot.fromJson(Map<String, dynamic> json) => MetricsSnapshot(
    timestamp: json['timestamp'] as int? ?? 0,
    system: SystemInfo.fromJson(json['system'] as Map<String, dynamic>? ?? {}),
    cpu: CpuMetrics.fromJson(json['cpu'] as Map<String, dynamic>? ?? {}),
    memory: MemoryMetrics.fromJson(json['memory'] as Map<String, dynamic>? ?? {}),
    disk: DiskMetrics.fromJson(json['disk'] as Map<String, dynamic>? ?? {}),
    network: NetworkMetrics.fromJson(json['network'] as Map<String, dynamic>? ?? {}),
  );
}
